### PR TITLE
fix softmax cast

### PR DESCRIPTION
**Pull request by fandaoyi**
fix softmax cast error

### CODE REVIEW

File: dipu/torch_dipu/dipu/device.py [hunk 1]
```diff
@@ -64,7 +64,7 @@ def __new__(cls, *args, **kwargs):
         return _device
 
 
-# always patch
+# always patch, device class is immuable, cannot directly patch __new__ method on python layer。
```
Comment (mrdanielw) [inline dipu/torch_dipu/dipu/device.py:67]: what's "immuable"？ typo?

File: dipu/torch_dipu/csrc_dipu/aten/ops/DIPUAmp.cpp
Comment (lljbash) [inline dipu/torch_dipu/csrc_dipu/aten/ops/DIPUAmp.cpp:1]: pytorch 2.1 把 amp 的实现搞清爽了很多，之后如果 dipu 支持 2.1 了我是打算把这里的代码也跟着改一下的（说不定不改也编不过），要不加个修改列表记录一下我们修改过什么东西。

File: dipu/torch_dipu/csrc_dipu/aten/ops/DIPUAmp.cpp [hunk 2]
```diff
@@ -88,8 +96,11 @@ template <DeviceType device_type, class Redispatch, Redispatch *F, class Ret,
 struct WrapFunction_<CastPolicy::fp32_set_opt_dtype, device_type, Redispatch, F,
                      Ret, guts::typelist::typelist<Args...>> {
   static Ret call(Args... args) {
-    c10::impl::ExcludeDispatchKeyGuard no_autocast(DispatchKey::Autocast);
-    if (firstarg_is_eligible(args...)) {
+    // DispatchKey::Autocast is not the alias key of all AutocastType as Autograd, 
+    // it's just alias of AutocastCUDA (see c10/core/DispatchKey.h)
```
Comment (lljbash) [inline dipu/torch_dipu/csrc_dipu/aten/ops/DIPUAmp.cpp:100]: 为啥以前测的 op 没问题，softmax 会挂？

File: dipu/torch_dipu/csrc_dipu/aten/ops/DIPUAmp.cpp [hunk 3]
```diff
@@ -24,6 +24,14 @@ namespace autocast {
...
 namespace {
 
+template <typename... Args>
+inline bool firstarg_is_eligible(
+    DeviceType device_type,
+    const Tensor& arg,
+    Args... args) {
+  return is_eligible(arg, device_type);
+}
+
```
Comment (mrdanielw) [inline dipu/torch_dipu/csrc_dipu/aten/ops/DIPUAmp.cpp:34]: 可以直接用pytorch的吗？   https://github.com/pytorch/pytorch/blob/a4391f085bff409dca93a8b3eff8e379f0ef8f68/aten/src/ATen/autocast_mode.h#L292
template <typename... Args>
inline bool firstarg_is_eligible(
    c10::DeviceType device_type,
    const Tensor& arg,
    Args... args) {
  return is_eligible(arg, device_type);
}

Comment (lljbash) [reply dipu/torch_dipu/csrc_dipu/aten/ops/DIPUAmp.cpp:34]: @mrdanielw 这个是 pytorch 2.1 的

File: dipu/tests/python/individual_scripts/test_op_benchmark.py [hunk 4]
```diff
@@ -22,7 +22,7 @@ def batched_dot_mul_sum(a, b):
 print(r0)
 # TODO(fandaoyi,lljbash): find out why it gets slower
 # assert r0.mean < 8.8e-5
-assert r0.mean < 18.8e-5
+assert r0.mean < 25.0e-5
```
[the remark this answers is not in the document]
Comment (lljbash) [reply dipu/tests/python/individual_scripts/test_op_benchmark.py:25]: 250 也不够，我刚刚 cuda 跑出一个 260

File: dipu/torch_dipu/dipu/device.py (hunk 1 above)
Comment (wey-code) [inline dipu/torch_dipu/dipu/device.py:67]:
```suggestion
# always patch, device class is immuable, cannot directly patch __new__ method on python layer.
```

File: dipu/torch_dipu/csrc_dipu/aten/ops/DIPUAmp.cpp (hunk 2 above)
Comment (wey-code) [reply dipu/torch_dipu/csrc_dipu/aten/ops/DIPUAmp.cpp:100]: 我们这个autocast的测例是不是每一种策略都需要加一下？

File: dipu/torch_dipu/dipu/device.py (hunk 1 above)
Comment (lljbash) [reply dipu/torch_dipu/dipu/device.py:67]:
```suggestion
# always patch, device class is immutable, cannot directly patch __new__ method on python layer.
```
immuable 是法语，daoyi tql

Comment (lljbash) [reply dipu/torch_dipu/dipu/device.py:67]: its French!

Comment (fandaoyi) [reply dipu/torch_dipu/dipu/device.py:67]: 额， 我改下吧，少些个 t

File: dipu/torch_dipu/csrc_dipu/aten/ops/DIPUAmp.cpp
Comment (fandaoyi) [reply dipu/torch_dipu/csrc_dipu/aten/ops/DIPUAmp.cpp:1]: 先别动了， 2.1 要考虑的很多， 还要看下是否要同时兼容， 下个版本单独建task 看吧。

File: dipu/torch_dipu/csrc_dipu/aten/ops/DIPUAmp.cpp (hunk 2 above)
Comment (fandaoyi) [reply dipu/torch_dipu/csrc_dipu/aten/ops/DIPUAmp.cpp:100]: 我挪到第一个 WrapFunction_ 重载处了， 没必要写那么多这种注释。

Comment (fandaoyi) [reply dipu/torch_dipu/csrc_dipu/aten/ops/DIPUAmp.cpp:100]: > 为啥以前测的 op 没问题，softmax 会挂？

别的不是那种写法。

File: dipu/torch_dipu/csrc_dipu/aten/ops/DIPUAmp.cpp (hunk 3 above)
Comment (fandaoyi) [reply dipu/torch_dipu/csrc_dipu/aten/ops/DIPUAmp.cpp:34]: 同上

File: dipu/tests/python/individual_scripts/test_op_benchmark.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (fandaoyi) [reply dipu/tests/python/individual_scripts/test_op_benchmark.py:25]: 我再改大点， 加个 warmup， cuda 再不行重试吧。这块后面再处理，这个benchmark  太粗糙。

**Review by mrdanielw — APPROVED** [review]

**Review by lljbash — APPROVED** [review]
